### PR TITLE
 Counter initial proposal - error handling

### PR DESCRIPTION
**Pull request by nieznanysprawiciel**
resolves: #402

Notes:
- Reject proposals with wrong generated ProposalId is not tested. It needs some work on our test utils.
- Probably we should do more security checks in negotiation API handlers, but this seems to be in another milestone.

### CODE REVIEW

**Review by tworec — COMMENTED** [review]
partial review

File: core/market/decentralized/src/negotiation/common.rs [hunk 1]
```diff
@@ -34,36 +38,66 @@ impl CommonBroker {
...
                 prev_proposal_id.clone(),
                 subscription_id.clone(),
             ))?
         }
 
+        if prev_proposal.body.issuer == IssuerType::Us {
+            return Err(ProposalError::OwnProposal(prev_proposal_id.clone()));
+        }
+
         let is_initial = prev_proposal.body.prev_proposal_id.is_none();
```
[the remark this answers is not in the document]
Comment (nieznanysprawiciel) [reply core/market/decentralized/src/negotiation/common.rs:87]: I will change `is_initial` name to something better

File: core/market/decentralized/src/negotiation/common.rs [hunk 2]
```diff
@@ -34,36 +38,66 @@ impl CommonBroker {
...
+                })?;
+        } else {
+            self.store
+                .get_demand(&subscription_id)
+                .await
+                .map_err(|e| match e {
+                    DemandError::NotFound(id) => ProposalError::NoSubscription(id),
+                    _ => ProposalError::InternalError(prev_proposal_id.clone(), e.to_string()),
+                })?;
+        }
```
Comment (nieznanysprawiciel) [inline core/market/decentralized/src/negotiation/common.rs:69]: Check if I can get proposal from database and use validate_subscription

File: core/market/decentralized/src/negotiation/common.rs [hunk 3]
```diff
@@ -153,25 +190,15 @@ impl CommonBroker {
...
         // TODO: do auth
         let _owner_id = NodeId::from_str(&caller)
             .map_err(|e| RemoteProposalError::Unexpected(e.to_string()))?;
 
-        let proposal = prev_proposal.from_draft(msg.proposal);
+        self.validate_subscription(&prev_proposal, owner).await?;
+        validate_match(&proposal, &prev_proposal).map_err(RemoteProposalError::NotMatching)?;
 
         self.db
             .as_dao::<ProposalDao>()
```
Comment (nieznanysprawiciel) [inline core/market/decentralized/src/negotiation/common.rs:204]: Consider mapping error to already countered

File: core/market/decentralized/src/negotiation/common.rs (hunk 2 above)
Comment (nieznanysprawiciel) [reply core/market/decentralized/src/negotiation/common.rs:69]: Moved to: https://github.com/golemfactory/yagna/pull/505
Affects to much code to do this in this PR.
Some tests won't pass until test utils for breaking networking will be merged.

File: core/market/decentralized/src/testing/events_helper.rs [hunk 4]
```diff
@@ -30,6 +45,18 @@ pub mod provider {
             _ => anyhow::bail!("Invalid event Type. ProposalEvent expected"),
         })
     }
+
+    pub async fn query_proposal(
+        market: &Arc<MarketService>,
+        offer_id: &SubscriptionId,
+        i: u8,
```
Comment (tworec) [inline core/market/decentralized/src/testing/events_helper.rs:52]: the name of this arg is my fault, but I'm surprised that you've not changed it ;)

File: core/market/decentralized/src/testing/mock_net.rs [hunk 5]
```diff
@@ -55,7 +55,9 @@ impl MockNet {
         };
 
         let mut inner = self.inner.lock().unwrap();
-        inner.nodes.insert(node_id.clone(), prefix);
+        if let Some(_) = inner.nodes.insert(node_id.clone(), prefix) {
+            panic!("[MockNet] Node [{}] already existed.", &node_id);
```
Comment (tworec) [inline core/market/decentralized/src/testing/mock_net.rs:59]: nooice 👍

File: core/market/decentralized/src/testing/mock_offer.rs [hunk 6]
```diff
@@ -70,6 +70,25 @@ pub mod client {
...
+                "golem": {
+                    "node.id.name": "its-test-provider",
+                    "node.debug.subnet": "blaa",
+                    "com.pricing.model": "linear"
+                },
+            }),
+            constraints![
+                "golem.node.debug.subnet" == "blaa",
+                "golem.srv.comp.expiration" > 0,
+                "custom.dontmatch" == "true",
```
Comment (tworec) [inline core/market/decentralized/src/testing/mock_offer.rs:86]: not a very subtle property name. ;))

File: core/market/decentralized/src/testing/events_helper.rs [hunk 7]
```diff
@@ -16,11 +19,23 @@ pub mod requestor {
             _ => anyhow::bail!("Invalid event Type. ProposalEvent expected"),
         })
     }
+
+    pub async fn query_proposal(
```
Comment (tworec) [inline core/market/decentralized/src/testing/events_helper.rs:23]: Maybe we can go even further, and define a trait for MarketService? Non-blocking comment

File: core/market/decentralized/tests/test_agreement.rs [hunk 8]
```diff
@@ -21,7 +23,9 @@ async fn test_gsb_get_agreement() -> Result<()> {
         .add_market_instance(PROV_NAME)
         .await?;
 
-    let proposal_id = exchange_draft_proposals(&network, REQ_NAME, PROV_NAME).await?;
+    let proposal_id = exchange_draft_proposals(&network, REQ_NAME, PROV_NAME)
```
Comment (tworec) [inline core/market/decentralized/tests/test_agreement.rs:26]: I was fighting to make it one-liner... 😢 
Non-blocking of course

File: core/market/decentralized/tests/test_agreement.rs [hunk 9]
```diff
@@ -603,7 +657,11 @@ async fn net_err_while_approving() -> Result<()> {
     let result = network
         .get_market(PROV_NAME)
         .provider_engine
-        .approve_agreement(prov_id.clone(), &agreement_id, 0.1)
+        .approve_agreement(
+            prov_id.clone(),
+            &agreement_id.clone().translate(OwnerType::Provider),
```
Comment (tworec) [inline core/market/decentralized/tests/test_agreement.rs:662]: I'd love to see the test for not translating `agreement_id`.  Does it make sense?

File: core/market/decentralized/src/testing/proposal_util.rs [hunk 10]
```diff
@@ -22,11 +31,7 @@ pub async fn exchange_draft_proposals(
     let offer_id = prov_mkt.subscribe_offer(&sample_offer(), &prov_id).await?;
 
     // Expect events generated on requestor market.
-    let req_events = req_mkt
-        .requestor_engine
-        .query_events(&demand_id, 3.14, Some(5))
-        .await?;
-    let req_offer_proposal1 = requestor::expect_proposal(req_events, 1)?;
+    let req_offer_proposal1 = requestor::query_proposal(&req_mkt, &demand_id, 1).await?;
```
Comment (tworec) [inline core/market/decentralized/src/testing/proposal_util.rs:34]: yeah, I feel much better when we are able to express more with less! 💪

File: core/market/decentralized/tests/test_negotiations.rs [hunk 11]
```diff
@@ -68,7 +66,8 @@ async fn test_exchanging_draft_proposals() -> Result<(), anyhow::Error> {
         Some(identity1.identity.to_string()),
     );
     assert_eq!(proposal1_prov.state, Some(State::Draft));
-    assert_eq!(proposal1_prov.prev_proposal_id, None);
+    // Requestor and Provider have different initial Proposals IDs
```
Comment (tworec) [inline core/market/decentralized/tests/test_negotiations.rs:69]: This is not the initial proposal. This is the first one seen on the Provider side.
```suggestion
    // Requestor and Provider have different first Proposals IDs
```

File: core/market/decentralized/tests/test_negotiations.rs [hunk 12]
```diff
@@ -140,3 +131,447 @@ async fn test_exchanging_draft_proposals() -> Result<(), anyhow::Error> {
...
+    // REQUESTOR side.
+    let proposal0 = requestor::query_proposal(&market1, &demand_id, 1).await?;
+    let proposal0_id = proposal0.get_proposal_id()?;
+
+    // Counter proposal for the first time.
+    let proposal1 = proposal0.counter_demand(sample_demand())?;
+    market1
+        .requestor_engine
+        .counter_proposal(&demand_id, &proposal0.get_proposal_id()?, &proposal1)
+        .await?;
```
Comment (tworec) [inline core/market/decentralized/tests/test_negotiations.rs:167]: I'd add here some basic assertion to check if the first attempt was successful. Make sense?

File: core/market/decentralized/tests/test_negotiations.rs [hunk 13]
```diff
@@ -140,3 +131,447 @@ async fn test_exchanging_draft_proposals() -> Result<(), anyhow::Error> {
...
+    // PROVIDER side
+    let proposal0 = provider::query_proposal(&market2, &offer_id, 2).await?;
+    let proposal0_id = proposal0.get_proposal_id()?;
+
+    // Counter proposal for the first time.
+    let proposal1 = proposal0.counter_offer(sample_offer())?;
+    market2
+        .provider_engine
+        .counter_proposal(&offer_id, &proposal0.get_proposal_id()?, &proposal1)
+        .await?;
```
Comment (tworec) [inline core/market/decentralized/tests/test_negotiations.rs:191]: ditto

File: core/market/decentralized/tests/test_negotiations.rs [hunk 14]
```diff
@@ -140,3 +131,447 @@ async fn test_exchanging_draft_proposals() -> Result<(), anyhow::Error> {
...
+    assert!(result.is_err());
+    match result.err().unwrap() {
+        ProposalError::OwnProposal(id) => assert_eq!(id, proposal1_id),
+        _ => panic!("Expected ProposalError::OwnProposal."),
+    }
+
+    Ok(())
+}
+
+/// Can't counter Proposal, for which Demand was unsubscribed.
```
Comment (tworec) [inline core/market/decentralized/tests/test_negotiations.rs:282]:
```suggestion
/// Requestor can't counter Proposal, for which he has unsubscribed Demand.
```

File: core/market/decentralized/tests/test_negotiations.rs [hunk 15]
```diff
@@ -140,3 +131,447 @@ async fn test_exchanging_draft_proposals() -> Result<(), anyhow::Error> {
...
+    assert!(result.is_err());
+    match result.err().unwrap() {
+        ProposalError::NoSubscription(id) => assert_eq!(id, demand_id),
+        _ => panic!("Expected ProposalError::Unsubscribed."),
+    }
+
+    Ok(())
+}
+
+/// Can't counter Proposal, for which Offer was unsubscribed.
```
Comment (tworec) [inline core/market/decentralized/tests/test_negotiations.rs:322]:
```suggestion
/// Provider can't counter Proposal, for which he has unsubscribed Offer.
```

File: core/market/decentralized/tests/test_negotiations.rs [hunk 16]
```diff
@@ -140,3 +131,447 @@ async fn test_exchanging_draft_proposals() -> Result<(), anyhow::Error> {
...
+
+    let demand_id = market1
+        .subscribe_demand(&sample_demand(), &identity1)
+        .await?;
+    let offer_id = market2.subscribe_offer(&sample_offer(), &identity2).await?;
+
+    let proposal0 = requestor::query_proposal(&market1, &demand_id, 1).await?;
+
+    // When we will counter this Proposal, Provider will have it already unsubscribed.
+    market2.unsubscribe_offer(&offer_id, &identity2).await?;
```
Comment (tworec) [inline core/market/decentralized/tests/test_negotiations.rs:396]: what about the switching order of those two instructions. How the quuery_proposal will behave on the Requestor side?

File: core/market/decentralized/tests/test_negotiations.rs [hunk 17]
```diff
@@ -140,3 +131,447 @@ async fn test_exchanging_draft_proposals() -> Result<(), anyhow::Error> {
...
+        offer_id,
+        demand_id,
+    } = exchange_draft_proposals(&network, "Node-1", "Node-2").await?;
+
+    let market1 = network.get_market("Node-1");
+    let market2 = network.get_market("Node-2");
+    let identity2 = network.get_default_id("Node-2");
+
+    // When we will counter this Proposal, Provider will have it already unsubscribed.
+    market2.unsubscribe_offer(&offer_id, &identity2).await?;
```
Comment (tworec) [inline core/market/decentralized/tests/test_negotiations.rs:437]: again, how do we behave when Offer unsubscription (on Provider) will occur before Requestor queries for Proposals?

File: core/market/resolver/src/resolver/matching.rs [hunk 18]
```diff
@@ -21,14 +21,14 @@ pub fn match_weak<'a>(
     demand: &'a PreparedDemand,
     offer: &'a PreparedOffer,
 ) -> Result<MatchResult<'a>, MatchError> {
-    log::debug!("Demand: {:?}", demand);
-    log::debug!("Offer: {:?}", offer);
+    log::trace!("Demand: {:?}", demand);
```
Comment (tworec) [inline core/market/resolver/src/resolver/matching.rs:24]: 👍

**Review by tworec — APPROVED** [review]
just minor comments

File: core/market/decentralized/tests/test_agreement.rs (hunk 9 above)
Comment (nieznanysprawiciel) [reply core/market/decentralized/tests/test_agreement.rs:662]: #510

File: core/market/decentralized/tests/test_negotiations.rs (hunk 17 above)
Comment (nieznanysprawiciel) [reply core/market/decentralized/tests/test_negotiations.rs:437]: #510

File: core/market/decentralized/tests/test_negotiations.rs (hunk 16 above)
Comment (nieznanysprawiciel) [reply core/market/decentralized/tests/test_negotiations.rs:396]: #510

**Review by tworec — APPROVED** [review]

**Pull request merged**